sklearn/metrics/_parallel_reductions.pyx

```diff
@@ -0,0 +1,1616 @@
+    # @final
+    cdef void compute_exact_distances(self) nogil:
+        """Convert proxy distances to pairwise distances in parallel."""
+        cdef:
+            ITYPE_t i, j
+
+        for i in prange(self.n_X, nogil=True, schedule='static',
+                        num_threads=self.effective_omp_n_thread):
+            for j in range(deref(self.neigh_indices)[i].size()):
+                deref(self.neigh_distances)[i][j] = (
+                        self.datasets_pair.distance_metric._rdist_to_dist(
+                            deref(self.neigh_distances)[i][j]
+                        )
+                )
+
+    @final
+    def compute(self,
+        str strategy = "auto",
+        bint return_distance = False,
+        bint sort_results = False
+    ):
+        if sort_results and not return_distance:
+            raise ValueError("return_distance must be True "
+                             "if sort_results is True.")
+
+        # Temporary datastructures which will be coerced to
+        # numpy arrays on return and then freed.
+        self.neigh_indices = new vector[vector[ITYPE_t]](self.n_X)
+        self.neigh_distances = new vector[vector[DTYPE_t]](self.n_X)
+
+        self.sort_results = sort_results
+
+        if strategy == 'auto':
+            # This is a simple heuristic whose constant for the
+            # comparison has been chosen based on experiments.
+            if 4 * self.chunk_size * self.effective_omp_n_thread < self.n_X:
+                strategy = 'parallel_on_X'
+            else:
+                strategy = 'parallel_on_Y'
+
+        # Limit the number of threads in second level of nested parallelism for BLAS
+        # to avoid threads over-subscription (in GEMM for instance).
+        with threadpool_limits(limits=1, user_api="blas"):
+            if strategy == 'parallel_on_Y':
+                self._parallel_on_Y()
+            elif strategy == 'parallel_on_X':
+                self._parallel_on_X()
+            else:
+                raise RuntimeError(f"strategy '{strategy}' not supported.")
+
+        if return_distance:
+            self.compute_exact_distances()
+            res = (
+                _coerce_vectors_to_np_nd_arrays(self.neigh_distances),
+                _coerce_vectors_to_np_nd_arrays(self.neigh_indices),
+            )
+        else:
+            res = _coerce_vectors_to_np_nd_arrays(self.neigh_indices)
+
+        del self.neigh_distances
+        del self.neigh_indices
+
+        return res
+
+
+cdef class FastSquaredEuclideanRadiusNeighborhood(RadiusNeighborhood):
+    """Fast specialized alternative for RadiusNeighborhood on EuclideanDistance.
+
+    Notes
+    -----
+    This implementation has an superior arithmetic intensity
+    and hence running time, but it can suffer from numerical
+    instability. RadiusNeighborhood with EuclideanDistance
+    must be used when exact precision is needed.
```

Comment:
```suggestion
    This implementation has a superior arithmetic intensity and hence
    better running time when the alternative is IO bound, but it can suffer
    from numerical instability.
    
    RadiusNeighborhood with EuclideanDistance must be used when higher
    numerical precision is needed.
```